src/core_deku/state.ml

```diff
@@ -101,113 +135,176 @@ let rec apply_user_operation ?sender t operation_hash user_operation =
-              ~get_contract_opt:(fun address ->
-                let address =
-                  Address.to_contract_hash address |> Option.value_exn in
-                Contract_storage.get_contract contract_storage ~address
-                |> Option.map ~f:(Fn.const (Address.of_contract_hash address)))
-          in
-          let%ok contract, user_op_list =
-            Contract_vm.Interpreter.invoke ~to_replace ~ctx ~arg:argument
-              ~gas:initial_gas contract
-            |> wrap_error in
+    Ledger.with_ticket_table t.ledger (fun ~get_table ~set_table ->
+        let table = get_table () in
+        let table_tickets, table =
+          Ticket_table.take_all_tickets table
+            ~sender:(Address.of_contract_hash to_invoke) in
+        let%ok _, table =
+          Ticket_table.take_tickets table ~sender
+            ~tickets:(List.map ~f:fst tickets) in
+        let ctx, to_replace =
+          let source = Address.of_key_hash source in
+          Context.make_state ~sender ~source
+            ~mapping:(Contract_vm.Contract.tickets_mapping contract)
+            ~contract_owned_tickets:table_tickets
+            ~self:(Address.of_contract_hash to_invoke)
+            ~provided_tickets:(Stdlib.List.to_seq tickets)
+            ~get_contract_opt:(fun address ->
+              let address =
+                Address.to_contract_hash address |> Option.value_exn in
+              Contract_storage.get_contract contract_storage ~address
+              |> Option.map ~f:(Fn.const (Address.of_contract_hash address)))
+        in
+        let%ok contract, user_op_list =
+          Contract_vm.Interpreter.invoke ~to_replace ~ctx ~arg:argument
+            ~gas:initial_gas contract
+          |> wrap_error in
 
-          let%ok tickets_mapping, tickets, operations =
-            ctx#finalize user_op_list
-            |> Result.map_error ~f:(fun _ ->
-                   `Invocation_error "execution error") in
-          let contract =
-            Contract_vm.Contract.update_tickets contract tickets_mapping in
-          let contract_storage =
-            Contract_storage.update_contract_storage contract_storage
-              ~address:to_invoke ~updated_contract:contract in
-          let table =
-            Ticket_table.update_tickets table
-              ~sender:(Address.of_contract_hash to_invoke)
-              ~tickets in
-          let ledger = set_table table in
-          let%ok t, receipt =
-            List.fold_result
-              ~f:(fun acc op ->
-                apply_contract_ops ~source
-                  ~sender:(Address.of_contract_hash to_invoke)
-                  acc op)
-              ~init:({ ledger; contract_storage }, None)
-              operations in
-          Ok (receipt, t)) in
-    Ok (t, receipt)
-
-and apply_contract_ops ~source ~sender (t, _) x =
+        let%ok tickets_mapping, tickets, operations =
+          ctx#finalize user_op_list
+          |> Result.map_error ~f:(fun _ ->
```

Comment:
What is the type of `_`? Could you change that to `()` or `_<name>` in order to make it more explicit?

Reply:
This was the receipt, but this code is overwritten now, so I don't get the comment